### PR TITLE
tanh float32 over 1.

### LINKED ISSUE #55390: tanh(float(7~8)) = 1.0000001 in XLA
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): linux or mac 
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: no
- TensorFlow installed from (source or binary): source 
- TensorFlow version (use command below): 2.8 or 1.15
- Python version: 3.9 or 2.7
- Bazel version (if compiling from source): bazel release 5.0.0-pre.20211011.2
- GCC/Compiler version (if compiling from source): Apple clang version 13.0.0 (clang-1300.0.29.3) 
- CUDA/cuDNN version: no
- GPU model and memory: no 

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`
    zsh: command not found: v1.12.1-68911-gea661077441

**Describe the current behavior**
The return value of tanh function is  1.0000001, but tanh can not return more than 1.

**Describe the expected behavior**
The return value of tanh function is 1 or less than 1.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): yes
- Briefly describe your candidate solution(if contributing):
There is a description in the xla comment that the xla implementation of the tanh function is copied from eigen3.(tensorflow/compiler/xla/service/llvm_ir/math_ops.h:25), but I compare the implementation of the tanh function within xla and eigen3, they not have same implementation .so I copy the implementation of tanh function from eigen3 to tensorflow ....

**Standalone code to reproduce the issue**
```
import os
os.environ['TF_XLA_FLAGS'] = '--tf_xla_auto_jit=2 --tf_xla_cpu_global_jit --tf_xla_min_cluster_size=0'

import tensorflow as tf

tf.compat.v1.disable_eager_execution()

#with tf.compat.v1.device('gpu'):
ctr_y = tf.compat.v1.random.uniform([1], minval=8, maxval=9, dtype=tf.compat.v1.float32)
ctr_pred_ori = tf.compat.v1.tanh(ctr_y)


session_config = tf.compat.v1.ConfigProto(allow_soft_placement=True, log_device_placement=True)
session_config.graph_options.rewrite_options.disable_meta_optimizer=True
with tf.compat.v1.Session(config=session_config) as sess:
  while  True:
    ret = sess.run([ctr_pred_ori])[0]
    if ret > 1.0:
      print(ret);
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
```
 212 [array([0.99999964], dtype=float32)]
13702 [array([0.99999976], dtype=float32)]
 621 [array([0.9999997], dtype=float32)]
17916 [array([0.9999998], dtype=float32)]
21312 [array([0.99999994], dtype=float32)]
48624 [array([0.9999999], dtype=float32)]
8097 [array([1.0000001], dtype=float32)]
   5 [array([1.0000002], dtype=float32)]
42322 [array([1.], dtype=float32)]
```
tanh return 1.0000001 8097 times

### PR DESCRIPTION
**Pull request by zhaozheng09**
issues: https://github.com/tensorflow/tensorflow/issues/55390 [-9.0, 9.0] is too precise causing tanh function overflow 
 in float32.

### CODE REVIEW

**Comment by zhaozheng09** [comment]
I got the correct answer under float type by XLA, but my change maybe  bring precision problem of double64 type, because the implementation of TANH in EIGEN3 is template<float>.

**Comment by gbaned** [comment]
@r4nt Can you please review this PR ? Thank you!

**Comment by zhaozheng09** [comment]
@gbaned

**Comment by gbaned** [comment]
> @gbaned

@zhaozheng09  Sorry for the delay. 
@r4nt  Can you please review this PR ? Thank you!

**Comment by zhaozheng09** [comment]
@gbaned emmm....please .

**Comment by cheshire** [comment]
Could you provide a test case which this fix resolves?

**Comment by zhaozheng09** [comment]
> Could you provide a test case which this fix resolves?

I'm sorry, I not find a example ...

**Comment by cheshire** [comment]
It does, search for RunAndCompare.

**Comment by zhaozheng09** [comment]
> It does, search for RunAndCompare.

Done, Check:
CPU float32 
CPU double64
GPU float32
GPU double64

**Comment by zhaozheng09** [comment]
@cheshire  please review~,thx.

File: tensorflow/compiler/xla/service/llvm_ir/math_ops.cc [hunk 1]
```diff
@@ -14,13 +14,19 @@ limitations under the License.
...
 
 namespace xla {
 namespace llvm_ir {
 
-llvm::Value* EmitFastTanh(llvm::IRBuilder<>* b, llvm::Value* input) {
+llvm::Value* EmitFastTanh(llvm::IRBuilder<>* b, llvm::Value* input,
+                          const TanhType input_type /* = TanhType::Double*/) {
   llvm::Type* type = input->getType();
+  const float plus_clamp =
+      (input_type == TanhType::Double) ? 9.0 : 7.99881172180175781f;
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/llvm_ir/math_ops.cc:27]: Why not derive `input_type` from input->Type() inside the function?

File: tensorflow/compiler/xla/service/llvm_ir/math_ops.cc [hunk 2]
```diff
@@ -14,13 +14,19 @@ limitations under the License.
...
+  const float minus_clamp =
+      (input_type == TanhType::Double) ? -9.0 : -7.99881172180175781f;
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/llvm_ir/math_ops.cc:29]: Move out constant into a `static constexpr`, reuse it for positive/negative case and give it a name? Also maybe a short comment on what values are chosen and why?

File: tensorflow/compiler/xla/tests/tanh_test.cc [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+namespace xla {
+namespace {
+
+class TanhTest : public ClientLibraryTestBase {
+ public:
+  ErrorSpec error_spec_{0.0001, 0.0001};
+  ErrorSpec strict_error_spec_{3.6e-15, 3.6e-15};
+};
+
+XLA_TEST_F(TanhTest, TanhF32sNormalTest) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/tests/tanh_test.cc:35]: How do these tests fail without your fix?

File: tensorflow/compiler/xla/tests/tanh_test.cc [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+#include "tensorflow/compiler/xla/types.h"
+#include "tensorflow/core/platform/types.h"
+
+namespace xla {
+namespace {
+
+class TanhTest : public ClientLibraryTestBase {
+ public:
+  ErrorSpec error_spec_{0.0001, 0.0001};
+  ErrorSpec strict_error_spec_{3.6e-15, 3.6e-15};
```
Comment (cheshire) [inline tensorflow/compiler/xla/tests/tanh_test.cc:32]: Error spec seems unused below?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/tests/tanh_test.cc (hunk 3 above)
Comment (zhaozheng09) [reply tensorflow/compiler/xla/tests/tanh_test.cc:35]: Old version will failed when tests are run .

File: tensorflow/compiler/xla/service/llvm_ir/math_ops.cc (hunk 2 above)
Comment (zhaozheng09) [reply tensorflow/compiler/xla/service/llvm_ir/math_ops.cc:29]: Origin version is 9.0, and I don't know why.

File: tensorflow/compiler/xla/service/llvm_ir/math_ops.cc (hunk 1 above)
Comment (zhaozheng09) [reply tensorflow/compiler/xla/service/llvm_ir/math_ops.cc:27]: ```input_type ``` maybe ```vector/array```, so I write a custom enum type.

**Comment by cheshire** [comment]
@rmlarsen

File: tensorflow/compiler/xla/tests/tanh_test.cc [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+class TanhTest : public ClientLibraryTestBase {};
+
+XLA_TEST_F(TanhTest, TanhF32sNormalTest) {
+  // This is like the test TanhTest.TanhF32s above, except that
+  // the input tensor is large enough to exercise the vectorized tanh
+  // implementation on XLA CPU.
+  XlaBuilder builder(TestName());
+  auto input_literal = LiteralUtil::CreateR1<float>(
+      {1.02,  -0.32, 0.85,  0.90,  1.23,  -0.91, -0.49, 0.80,  -0.67, 0.16,
```
Comment (cheshire) [inline tensorflow/compiler/xla/tests/tanh_test.cc:37]: We actually have exhaustive tanh f32 tests: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/tests/exhaustive_unary_test_f32_or_smaller.cc;l=453-463?q=tensorflow%2Fcompiler%2Fxla%2Ftests%2Fexhaustive_unary_test_f32_or_smaller.cc

could we tighten the error spec there and add required invariants instead?

File: tensorflow/compiler/xla/service/llvm_ir/math_ops.h [hunk 6]
```diff
@@ -22,9 +22,25 @@ limitations under the License.
 namespace xla {
 namespace llvm_ir {
 
+enum TanhType {
+  Double = 0,
+  Float = 1,
+};
+
+static constexpr float UPPER_BOUNDER_DOUBLE = 9.0;
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/llvm_ir/math_ops.h:30]: Constant code style is kUpperBoundDouble, probably we could use a more descriptive name? (like, specify tanh).

Does negation deserve it's own constant, or could we negate existing one?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/llvm_ir/math_ops.cc (hunk 1 above)
Comment (cheshire) [reply tensorflow/compiler/xla/service/llvm_ir/math_ops.cc]: Could we get input type from inside the vector?

**Comment by cheshire** [comment]
@gbaned the comment about explicit tests wasn't addressed.

File: tensorflow/compiler/xla/tests/tanh_test.cc (hunk 5 above)
Comment (zhaozheng09) [reply tensorflow/compiler/xla/tests/tanh_test.cc:37]: @x@ 
I don't know much about XLA, I don't think my single test is suitable for embedding ```Exhaustive32BitOrLessUnaryTest```.
1. ```Exhaustive32BitOrLessUnaryTest``` is a batch test, but bounder test is not need batch test.
2.  input of ```Exhaustive32BitOrLessUnaryTest``` must be int type, but I need a float or double input .

**Comment by cheshire** [comment]
> I don't know much about XLA, I don't think my single test is suitable for embedding Exhaustive32BitOrLessUnaryTest.
> input of Exhaustive32BitOrLessUnaryTest must be int type, but I need a float or double input .

Saw this in email, don't see it in GH UI somehow. It looks float to me, why does it need to be an int type?

File: tensorflow/compiler/xla/tests/tanh_test.cc (hunk 5 above)
Comment (zhaozheng09) [reply tensorflow/compiler/xla/tests/tanh_test.cc:37]: the unittest of tanh is not very suitable to be placed in this test.
1. the interface of ```run``` only recv a xla_op, I need ```Max(Tanh(input), input_boundary);``` to express ```output < 1```
2. I want to express ```output < 1```, but error_spec_gen only express ```expect - 0.0001 < output < expect + 0.0001```

Comment (cheshire) [reply tensorflow/compiler/xla/tests/tanh_test.cc:37]: Would it be possible to modify that test instead then? E.g. modify the `Run` method to accept an optional `check_invariant` callback, which will check the desired property (|output| < 1 in your case)? That seems better than having a hardcoded set of outputs.

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/tests/tanh_test.cc (hunk 5 above)
Comment (zhaozheng09) [reply tensorflow/compiler/xla/tests/tanh_test.cc]: done

**Comment by zhaozheng09** [comment]
@cheshire  please review, thx~

**Review by cheshire — APPROVED** [review]

**Comment by gbaned** [comment]
@zhaozheng09 Can you please resolve conflicts? Thank you!

**Review by cheshire — APPROVED** [review]

**Comment by zhaozheng09** [comment]
what  should I do now ? @gbaned

**Comment by gbaned** [comment]
> what should I do now ? @gbaned

Hi @zhaozheng09  Nothing pending your end now. We will le you know if anything needed. Thank you!

**Comment by cheshire** [comment]
Seems there is a test crash in `//tensorflow/compiler/xla/service/cpu/tests:cpu_intrinsic_test`:

```
[ RUN      ] CpuUnaryIntrinsicTestInstantiation/CpuUnaryIntrinsicTest.DoIt/Tanh_On_x86_64
2022-12-14 10:24:49.372274: E [tensorflow/compiler/xla/tests/filecheck.cc:69](https://cs.corp.google.com/piper///depot/google3/tensorflow/compiler/xla/tests/filecheck.cc?l=69&cl=494035788)] Tried to execute FileCheck at /b/f/w/bazel-out/k8-opt/bin/tensorflow/compiler/xla/service/cpu/tests/cpu_intrinsic_test.runfiles/org_tensorflow/external/llvm-project/llvm/FileCheck
2022-12-14 10:24:49.372412: E [tensorflow/compiler/xla/tests/filecheck.cc:75](https://cs.corp.google.com/piper///depot/google3/tensorflow/compiler/xla/tests/filecheck.cc?l=75&cl=494035788)] FileCheck stderr:
/[b/f/w/_tmp/ac029b7a7bbd9685034033fb3a65e865/tempfile-localhost-d2c2d32c-15-5efc72571126d:1](https://cs.corp.google.com/piper///depot/google3/b/f/w/_tmp/ac029b7a7bbd9685034033fb3a65e865/tempfile-localhost-d2c2d32c-15-5efc72571126d?l=1&cl=494035788):8: error: CHECK: expected string not found in input
CHECK: fcmp fast uge <4 x float> %wide.load, <float -9.000000e+00, float -9.000000e+00, float -9.000000e+00, float -9.000000e+00>
       ^
<stdin>:1:1: note: scanning from here
; ModuleID = '__compute_module'
^
<stdin>:22:7: note: possible intended match here
 %5 = fcmp fast uge <4 x float> %wide.load, <float 0xC01F9F09E0000000, float 0xC01F9F09E0000000, float 0xC01F9F09E0000000, float 0xC01F9F09E0000000>
      ^

Input file: <stdin>
Check file: /b/f/w/_tmp/ac029b7a7bbd9685034033fb3a65e865/tempfile-localhost-d2c2d32c-15-5efc72571126d

-dump-input=help explains the following input dump.

Input was:
<<<<<<
```

**Comment by zhaozheng09** [comment]
> Seems there is a test crash in `//tensorflow/compiler/xla/service/cpu/tests:cpu_intrinsic_test`:
> 
> ```
> [ RUN      ] CpuUnaryIntrinsicTestInstantiation/CpuUnaryIntrinsicTest.DoIt/Tanh_On_x86_64
> 2022-12-14 10:24:49.372274: E [tensorflow/compiler/xla/tests/filecheck.cc:69](https://cs.corp.google.com/piper///depot/google3/tensorflow/compiler/xla/tests/filecheck.cc?l=69&cl=494035788)] Tried to execute FileCheck at /b/f/w/bazel-out/k8-opt/bin/tensorflow/compiler/xla/service/cpu/tests/cpu_intrinsic_test.runfiles/org_tensorflow/external/llvm-project/llvm/FileCheck
> 2022-12-14 10:24:49.372412: E [tensorflow/compiler/xla/tests/filecheck.cc:75](https://cs.corp.google.com/piper///depot/google3/tensorflow/compiler/xla/tests/filecheck.cc?l=75&cl=494035788)] FileCheck stderr:
> /[b/f/w/_tmp/ac029b7a7bbd9685034033fb3a65e865/tempfile-localhost-d2c2d32c-15-5efc72571126d:1](https://cs.corp.google.com/piper///depot/google3/b/f/w/_tmp/ac029b7a7bbd9685034033fb3a65e865/tempfile-localhost-d2c2d32c-15-5efc72571126d?l=1&cl=494035788):8: error: CHECK: expected string not found in input
> CHECK: fcmp fast uge <4 x float> %wide.load, <float -9.000000e+00, float -9.000000e+00, float -9.000000e+00, float -9.000000e+00>
>        ^
> <stdin>:1:1: note: scanning from here
> ; ModuleID = '__compute_module'
> ^
> <stdin>:22:7: note: possible intended match here
>  %5 = fcmp fast uge <4 x float> %wide.load, <float 0xC01F9F09E0000000, float 0xC01F9F09E0000000, float 0xC01F9F09E0000000, float 0xC01F9F09E0000000>
>       ^
> 
> Input file: <stdin>
> Check file: /b/f/w/_tmp/ac029b7a7bbd9685034033fb3a65e865/tempfile-localhost-d2c2d32c-15-5efc72571126d
> 
> -dump-input=help explains the following input dump.
> 
> Input was:
> <<<<<<
> ```

Done

**Review by cheshire — APPROVED** [review]

File: tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h [hunk 7]
```diff
@@ -943,6 +952,23 @@ inline std::vector<std::pair<int64_t, int64_t>> CreateExhaustiveF32Ranges() {
   return result;
 }
 
+template <typename T>
+bool CheckValidRange(const T input, const T max, const T min) const {
```
Comment (akuegel) [inline tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h:956]: non-member function cannot have 'const' qualifier
It should be:

```
template <typename T>
bool CheckValidRange(const T input, const T max, const T min) {
```

File: tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h [hunk 8]
```diff
@@ -943,6 +952,23 @@ inline std::vector<std::pair<int64_t, int64_t>> CreateExhaustiveF32Ranges() {
...
+  return (input < max) && (input > min);
+}
+
+template <>
+bool CheckValidRange(const std::complex<double>, const std::complex<double>,
+                     const std::complex<double>) const {
```
Comment (akuegel) [inline tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h:962]: non-member function cannot have 'const' qualifier

File: tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h [hunk 9]
```diff
@@ -943,6 +952,23 @@ inline std::vector<std::pair<int64_t, int64_t>> CreateExhaustiveF32Ranges() {
...
+  return true;
+}
+
+template <>
+bool CheckValidRange(const std::complex<float>, const std::complex<float>,
+                     const std::complex<float>) const {
```
Comment (akuegel) [inline tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h:968]: non-member function cannot have 'const' qualifier

**Review by akuegel — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h (hunk 8 above)
Comment (cheshire) [reply tensorflow/compiler/xla/tests/exhaustive/exhaustive_op_test_utils.h]: How are these not caught by the Github CI btw?

**Review by akuegel — APPROVED** [review]

**Comment by akuegel** [comment]
So unfortunately this change got rolled back because some tests started failing, most likely too strict allowed deviations. I will try to get it rolled forward again at the beginning of next year when I am back from vacation.

**Comment by zhaozheng09** [comment]
@akuegel  Sorry for the extra work for you. Can you give me a small example to reproduce this problem? 
I'm wondering if my PR will affect performance? ```2984.331 us > 2836.245 us``` ?

**Comment by akuegel** [comment]
> > @akuegel  Sorry for the extra work for you. Can you give me a small example to reproduce this problem? 
> > I'm wondering if my PR will affect performance? ```2984.331 us > 2836.245 us``` ? 
> 
> Unfortunately it is an internal test, so I will have to adjust it. But I think your change is ok, just the epsilon value for the test needs to be adjusted. Sorry that I couldn't get this done before my vacation.

Also ignore those extra remarks, I think this is a bug that they are added to this PR. It seems currently every commit to the OpenXLA repository shows up here. @joker-eph is this a known issue?

**Comment by zhaozheng09** [comment]
@akuegel  

Happy holidays to you, as long as I know I don't have to do anything for me.

**Comment by zhaozheng09** [comment]
@gbaned What should I do now ?

**Comment by akuegel** [comment]
Unfortunately you can do nothing :-( This is still blocked by an internal test failure. There were two test failures, one of them I can fix myself by increasing the allowed epsilon value for numerical differences. The other test is in a code base where I don't have access to, so I contacted the team to ask them whether they can check what can be done.

**Comment by zhaozheng09** [comment]
> Unfortunately you can do nothing :-( This is still blocked by an internal test failure. There were two test failures, one of them I can fix myself by increasing the allowed epsilon value for numerical differences. The other test is in a code base where I don't have access to, so I contacted the team to ask them whether they can check what can be done.

Thanks for your hard work.

**Pull request merged**